source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1235,6 +1235,122 @@ the pool's generation number then error handling MUST continue according to
+  MongoDB handshake and completing authentication (if configured).
+- A network error or network timeout may occur while reading or writing to an
+  established connection.
+- A command error may be returned from the server.
+- A "writeConcernError" field may be included in the command response.
+
+Depending on the context, these errors may update update SDAM state by marking
+the server Unknown and may clear the server's connection pool. Some errors
+also require other side effects, like cancelling a check or requesting an
+immediate check. Drivers may use the following pseudocode to guide their
+implementation::
+
+    def handleError(error):
+        address = error.address
+        topologyVersion = error.topologyVersion
+
+        with client.lock:
+            # Ignore stale errors based on generation and topologyVersion.
+            if isStaleError(client.topologyDescription, error)
+                return
+
+            if isStateChangeError(error):
+                # Mark the server Unknown
+                unknown = new ServerDescription(type=Unknown, error=error, topologyVersion=topologyVersion)
+                onServerDescriptionChanged(unknown, connection pool for server)
+                if isShutdown(code) or (error was from <4.2):
+                  # the pools must only be cleared while the lock is held.
+                  clear connection pool for server
+                if multi-threaded:
+                    request immediate check
+                else:
+                    # Check right now if this is "not master", since it might be a
+                    # useful secondary. If it's "node is recovering" leave it for the
+                    # next full scan.
+                    if isNotMaster(error):
+                        check failing server
+            elif isNetworkError(error) or (not error.completedHandshake and (isNetworkTimeout(error) or isAuthError(error))):
+                # Mark the server Unknown
+                unknown = new ServerDescription(type=Unknown, error=error)
+                onServerDescriptionChanged(unknown, connection pool for server)
+                clear connection pool for server
+                # Cancel inprogress check
+                cancel monitor check
+
+    def isStaleError(topologyDescription, error):
+        currentServer = topologyDescription.servers[server.address]
+        currentGeneration = currentServer.pool.generation
+        generation = get connection generation from error
+        if generation < currentGeneration:
+            # Stale generation number.
+            return True
+
+        currentTopologyVersion = currentServer.topologyVersion
+        # True if the current error's topologyVersion is greater than the server's
+        # We use >= instead of > because any state change should result in a new topologyVersion
+        return compareTopologyVersion(currentTopologyVersion, error.commandResponse.get("topologyVersion")) >= 0
```

Comment:
What does `compareTopologyVersion` do on the inside? Is it not a simple number comparison?

Reply:
It's defined here: https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#replacing-the-topologydescription

Note I didn't change isStaleError at all in this pr, I only shuffled the code from down below.